### DISCUSSION TITLE
Downtime during migrations

### DISCUSSION THREAD

**Discussion by tevintchuinkam** [pocketbase/pocketbase#5678] [Q&A] [answered]
I recently read [this great article](https://blog.wesleyac.com/posts/consider-sqlite) on how why SQLite is a great choice for web apps. 

One of the only drawbacks was the following: 

_"""SQLite has essentially no support for live migrations, so you need to instead make a new table, copy the data from the old table into the new one, and switch over. This is slow if you have a very large number of rows, and quite difficult to do without downtime in general. For most applications, it's fine to take a few seconds or a minute of downtime when doing a complex migration, but if you have many gigabytes or more of data that you want to migrate, it's worth looking carefully at how difficult that'll be"""._

The article is 3 years old. Could someone comment of this is still something to consider and how this affects pocketbase ? 
If it does, has anyone done tests on how much downtime one has to expect when applying migrations on a Pocketbase DB that contains multiple GB of data?
Does Pocketbase have some sort of buffer to hold incoming queries that can't be executed right away ? Or might it be useful to use a message broker like RabbitMQ to deal with this!

Greatful for any insights! Thanks 🙏

**Comment by ganigeorgiev** [comment, marked as answer]
It depends what do you understand by "live migrations" and what the implied migrations do in the first place.

SQLite supports limited set of `ALTER` statements like renaming a table, renaming/adding/dropping columns, adding/removing indexes, etc. without having to create a separate temp structure during schema changes and based on my experience for most of the time it is more than enough.

PocketBase maintains 2 different pools of connection:
- reads pool with max 120 open connections (can be adjusted via the `pocketbase.Config` struct during initialization)
- write pool with max 1 open connection

For most of the time readers and writer don't block each other (_there are some exceptions when dealing with mixed read&write transactions_), but even when they block we have auto fail/retry handling on application level in addition to the builtin SQLite `busy_timeout` polling. Blocked requests will simply queue up - not indefinitely but long enough to not affect significantly regular application usage for most of the time.

When performing upgrades from one PocketBase version to another there will be always downtime when restarting the process but Go startup time is almost instantaneous and very rarely a compiled Go application takes more than 1s to boot.

In any case I don't think write locks in this scenario are avoidable even with other traditional databases (if I'm not mistaken PostgreSQL uses by default exclusive table locks during schema changes).

But is this really a problem?

How often do you change the type definition of existing table column in production?

How much data do you store and operate with?
_Note that even 1GB database file is enormous amount of data that many users probably might never reach. The vacuumed size of my other open source project (Presentator) `pb_data/data.db` is ~150MB and this project is live since 2015/2016._

Is really the ~1s downtime when switching binaries not acceptable?

With that said, instead of relying on someone to tell you how "fast" and reliable something is, try to create a test setup and measure it on your own based on your app specific usage because often the only correct answer would be "it depends".
